Process all pages and don't Liquid-render pages corpus

mbland opened the pull request:
This is really two PRs in one, motivated by my work to build [18F/jekyll_pages_api_search](https://github.com/18F/jekyll_pages_api_search/). First, it adds Posts, Documents, and StaticFiles to the pages corpus. Second, it prevents the resulting `pages.json` corpus from being rendered by Liquid.

I noticed the need for these features when experimenting with interating the jekyll_pages_api_search plugin into [18F/18f.gsa.gov](https://github.com/18F/18f.gsa.gov/). The absence of Posts from the corpus made the search less useful; then the presence of example Liquid markup in the [Jekyll and webhooks post](https://18f.gsa.gov/2014/11/17/taking-control-of-our-website-with-jekyll-and-webhooks/) caused the `pages.json` file to be invalid JSON due to rendering the corpus with Liquid, rather than just assigning the JSON output directly to `page.output`.

With these changes, I'm able to build and integrate search indexes both into 18f.gsa.gov and the [18F/hub](https://github.com/18F/hub/).

cc: @afeld

mbland commented:
cc: @adelevie, since Aidan's mostly out this week.

adelevie commented:
LGTM, merging in 30 minutes.

mbland commented:
@adelevie Just added a "Developing" section to the README. If it looks sufficient, I'll add a similar section to 18F/jekyll_pages_api_search#2.

afeld commented on README.md:
````diff
@@ -40,6 +40,25 @@ You can then see the generated JSON file at http://localhost:4000/api/v1/pages.j
+group :jekyll_plugins do
+  gem 'jekyll_pages_api', :path => '../jekyll_pages_api'
+end
+```
````
Nice.

afeld commented on Rakefile:
```diff
@@ -1,2 +1,7 @@
+  require 'rspec/core/rake_task'
+    RSpec::Core::RakeTask.new(:spec)
+rescue LoadError
+end
```
I am personally fine with leaving the spec running separate from the Rake tasks...

mbland replied:
But does it hurt having the rake task in there for consistency with using rake to build and release the gem?

afeld replied:
Nope. No worries.

afeld commented on lib/jekyll_pages_api/generator.rb:
```diff
@@ -10,7 +12,12 @@ def initialize(site)
 
     def pages
-      self.site.pages.map{|page| Page.new(page) }.select(&:html?)
+      result = Array.new
```
Minor: why use this rather than an array literal?

mbland replied:
Oh, my paranoia again. Updated to use `[]`.

afeld commented on lib/jekyll_pages_api/generator.rb:
```diff
@@ -10,7 +12,12 @@ def initialize(site)
     def pages
-      self.site.pages.map{|page| Page.new(page) }.select(&:html?)
+      result = Array.new
+      self.site.each_site_file do |p|
```
Is this an internal Jekyll API? Seems like a good find, though it could be more fragile. We should keep an eye out for it changing.

mbland replied:
Yeah, I suppose it is. :-) This is the current implementation:

```ruby
    def each_site_file
      %w(posts pages static_files docs_to_write).each do |type|
        send(type).each do |item|
          yield item
        end
      end
    end
```
`posts`, `pages`, and `static_files` are listed with `attr_accessor`. `docs_to_write` is a method that runs through collection documents. We can define our own method instead of using `each_site_file`, or keep it for now and fix it if it ever changes, which shouldn't require major surgery.

afeld replied:
Fine by me!

afeld commented on lib/jekyll_pages_api/generator.rb:
```diff
@@ -10,7 +12,12 @@ def initialize(site)
-      self.site.pages.map{|page| Page.new(page) }.select(&:html?)
+      result = Array.new
+      self.site.each_site_file do |p|
+        p = Page.new(p, @site)
```
Tsk tsk single-letter variables :wink: Especially this one.

http://ruby-doc.org/core-2.2.2/Kernel.html#method-i-p

mbland replied:
More like tsk tsk single-letter method names...but updated with more descriptive names.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -3,14 +3,17 @@
 module JekyllPagesApi
   # wrapper for a Jekyll::Page
   class Page
+    HTML_EXTS = %w(.html .md .markdown .textile)
```
Minor: use a Set.

http://ruby-doc.org/stdlib-2.2.2/libdoc/set/rdoc/Set.html#method-i-to_set

mbland replied:
Wow, never knew about Set in Ruby. Shoulda thought to look for it. Done.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -18,23 +21,31 @@ def filterer
     def title
-      self.filterer.decode_html(self.page.data['title'] || '')
+      title = self.page.data['title'] if self.page.respond_to?(:data)
+      title = self.page.title if title == nil and self.page.respond_to?(:title)
```
You can simplify this line to

```ruby
title ||= self.page.title if self.page.respond_to?(:title)
```

Are the conditionals to handle Pages vs. Posts etc? Might be worth checking for `is_a?`, or if you want to stick with duck-typing, at least adding a comment of the types that would be expected for each.

mbland replied:
Did the code change but to your point, using `is_a?` won't currently work with the tests because Rspec doubles are used, and those don't appear to actually inherit from the original class. I'd experimented with re-opening `Jekyll::Page` et. al. to extend them with new methods, but it failed the tests for the same reason.

So, what I could do is create `DummyPage < Jekyll::Page`, et. al., that prevent the initializers from running, and use those instead of the Rspec-generated doubles. Then I'd prefer to do the class-extension trick rather than `is_a?` or `responds_to?`. Thoughts?

afeld replied:
Meh, not worth it.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -3,14 +3,17 @@
 
     def html?
-      %w(.html .md).include?(self.page.ext)
+      path = @page.path
```
Blech instance variables :wink: Mind using `self.page`?

mbland replied:
So I figure [this is the rationale for `self.page` vs. `@page`]? Done, but I fundamentally object to the premise that the `self.page` form is preferable because it encourages subclassing. ;-)

mbland replied:
Although, I guess it also enables one invoke a method on the same object should you decide to switch from a bare instance var to an accessor... OK, I've overruled my own objection. :-P

afeld replied:
My main gripe is that if you misspell them, they give back `nil` instead of raising an error, which has caused weird bugs for me in the past. Also a bonus that you can easily swap the reader out for a method without changes anywhere else.

mbland replied:
That's a very compelling argument. Had no idea about that `nil` behavior. You've probably just added a few hours or a couple of days back to my life, at least.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -3,14 +3,17 @@
-    def initialize(page)
+    def initialize(page, site)
       @page = page
+      @site = site
```
Why wouldn't we want to use the `page.site`?

mbland replied:
Because `Jekyll::StaticFile` doesn't implement that accessor.

afeld replied:
Maybe worth putting that in an inline comment?

mbland replied:
Done in #25.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -18,23 +21,31 @@ def filterer
+      rel_path = self.page.url if self.page.respond_to?(:url)
+      if rel_path == nil and self.page.respond_to?(:relative_path)
+        rel_path = self.page.relative_path
+      end
```
I'd simplify (and make more consistent with above) with

```ruby
rel_path ||= self.page.relative_path if self.page.respond_to?(:relative_path)
```

afeld replied:
Also, maybe worth moving these lines to a `rel_path` method.

mbland replied:
Done and done.

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -18,23 +21,31 @@ def filterer
     def body_text
-      self.filterer.text_only(self.page.content)
+      output = self.page.content if self.page.respond_to?(:content)
+      File.open(self.page.path, 'r') {|f| output = f.read} if output == nil
```
Simpler as

```ruby
output ||= File.read(self.page.path)
```

mbland replied:
Dammit, I keep expecting http://ruby-doc.org/core-2.2.0/ to have the _complete_ documentation; now I know to always check `ri`. Done.

afeld replied:
Well, [it does](http://ruby-doc.org/core-2.0.0/IO.html#method-c-read), but that method is inherited from the `IO` base class.

FYI that I like RubyDoc.info slightly better because it shows all the inherited methods.

http://www.rubydoc.info/stdlib/core/2.1.0/File

mbland replied:
Ah, thanks for that!

afeld commented on lib/jekyll_pages_api/version.rb:
```diff
@@ -1,3 +1,3 @@
 module JekyllPagesApi
-  VERSION = '0.1.1'
+  VERSION = '0.1.3'
```
Mind if we do this bump separately?

mbland replied:
Pulled it down to 0.1.2. That version is released, but somehow hasn't yet been committed.

afeld commented on spec/integration_spec.rb:
```diff
@@ -13,8 +13,7 @@ def read_json(path)
   def entries_data
-    json = read_json(JSON_PATH)
-    json['entries']
+    (@json ||= read_json(JSON_PATH))['entries']
```
Minor: I would move out to a separate method.

```ruby
def pages_data
  @pages_data ||= read_json(JSON_PATH)
end
```

mbland replied:
Done.

afeld commented on spec/page_spec.rb:
```diff
@@ -1,19 +1,151 @@
+
+BASEURL = Jekyll::Configuration::DEFAULTS['baseurl']
+
+RSpec.shared_context "create page" do
```
Ah cool, I didn't know about this RSpec feature!

mbland replied:
Me either. In fact, this is the first time I've ever used Rspec. :-)

afeld commented on spec/page_spec.rb:
```diff
@@ -1,19 +1,151 @@
+
+RSpec.shared_context "create page" do
+  before { @tmp_dir = '' }
+  after { FileUtils.remove_entry(@tmp_dir) if not @tmp_dir.empty? }
```
Minor: simpler as

```ruby
before { @tmp_dir = nil }
after { FileUtils.remove_entry(@tmp_dir) if @tmp_dir }
```

mbland replied:
Done.

afeld commented on spec/page_spec.rb:
```diff
@@ -1,19 +1,151 @@
+    allow(jekyll_doc).to receive(:data).and_return(data) unless data == nil
+    JekyllPagesApi::Page.new(jekyll_doc, site)
+  end
+end
```
Can we move this to its own file in `spec/support/`?

mbland replied:
Done.

afeld commented:
Not sure why I was so nitpicky on this one, sorry! Looks really good overall. A few overall comments:

* We might want to add something to the README saying what types of Documents will and won't be included in the API.
* Not sure about including all Document types in the `pages.json` (/cc https://github.com/18F/jekyll_pages_api/issues/2)...thoughts?
* FYI, a Rubyism is to use `something.nil?` rather than `something == nil`. Not sure why.

Thanks!

mbland commented:
OK @afeld, just push a commit addressing all your comments from above. I updated the README to mention that all Pages, Posts, Documents and StaticFiles are processed. It's possible we may want to leave out StaticFiles, but if they happen to contain HTML, why not include them? Also, I like the `.nil?` notation; very convenient and compact. As always, I learn all these little idioms that make me a better Rubyist from your reviews. :-)

afeld commented on lib/jekyll_pages_api/page.rb:
```diff
@@ -18,23 +21,33 @@ def filterer
 
     def base_url
-      self.page.site.baseurl
+      @site.baseurl
```
Minor: set up an `attr_reader` for `:site` and use `self.site`.

mbland replied:
Done in #25.
